hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,239 @@
+        new ECContainerReplicaCount(container, replicas, pendingOps,
+            remainingMaintenanceRedundancy);
+
+    ContainerHealthResult currentUnderRepRes = ecContainerHealthCheck
+        .checkHealth(container, replicas, pendingOps,
+            remainingMaintenanceRedundancy);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+    if (currentUnderRepRes
+        .getHealthState() != ContainerHealthResult.HealthState
+        .UNDER_REPLICATED) {
+      LOG.info("The container {} state changed and it's not in under"
+              + " replication any more. Current state is: {}",
+          container.getContainerID(), currentUnderRepRes);
+      return emptyMap();
+    }
+    ContainerHealthResult.UnderReplicatedHealthResult containerHealthResult =
+        ((ContainerHealthResult.UnderReplicatedHealthResult)
+            currentUnderRepRes);
+    if (containerHealthResult.isSufficientlyReplicatedAfterPending()) {
+      LOG.info("The container {} with replicas {} is sufficiently replicated",
+          container.getContainerID(), replicaCount.getReplicas());
+      return emptyMap();
+    }
+    if (replicaCount.unRecoverable()) {
+      LOG.warn("The container {} is unrecoverable. The available replicas" +
+          " are: {}.", container.containerID(), replicaCount.getReplicas());
+      return emptyMap();
+    }
+    final ContainerID id = container.containerID();
+    final Map<DatanodeDetails, SCMCommand<?>> commands = new HashMap<>();
+    try {
+     // State is UNDER_REPLICATED
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOps) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      // We got the missing indexes, this is excluded any decommissioning
+      // indexes. Find the good source nodes.
+      if (missingIndexes.size() > 0 && !replicaCount.unRecoverable()) {
+        List<ContainerReplica> sources = replicas.stream().filter(r -> r
+            .getState() == StorageContainerDatanodeProtocolProtos
+            .ContainerReplicaProto.State.CLOSED)
+            // Exclude stale and dead nodes. This is particularly important for
+            // maintenance nodes, as the replicas will remain present in the
+            // container manager, even when they go dead.
+            .filter(r -> ReplicationManager
+                .getNodeStatus(r.getDatanodeDetails(), nodeManager).isHealthy())
+            .filter(r -> !deletionInFlight.contains(r.getDatanodeDetails()))
+            .collect(Collectors.toList());
+        LOG.debug("Missing indexes detected for the container {}." +
+                " The missing indexes are {}", id, missingIndexes);
+        // We have source nodes.
+        if (sources.size() >= repConfig.getData()) {
```

Comment:
we should not just check the size of sources here, we should check that the eligible source datanodes holds the different replica indexes , which should be larger than repConfig.getData()

Reply:
The API defined with replica set. However, ContainerReplica is not considered replica index in equals method. Even though it's set replicas can be duplicated if caller does not check. 

It will fail at DN side later anyway as not enough sources since we are putting them in map with replicaIndex as key:
SortedMap<Integer, DatanodeDetails> sourceNodeMap =
        reconstructionCommandInfo.getSources().stream().collect(Collectors
            .toMap(DatanodeDetailsAndReplicaIndex::getReplicaIndex,
                DatanodeDetailsAndReplicaIndex::getDnDetails, (v1, v2) -> v1,
                TreeMap::new));

But we can handle here itself. Let me think.

Reply:
Tried to map things into map with key ad replica index finally after filtering. So, the sources will be unique. 
Probably this filtering can be more based on DN load as well. Let's say if we have 2 src nodes with same replica index, that whichever node has less load that node can be chosen. That can be improved later.